### ISSUE TITLE
2.0.0 release plan

### ISSUE THREAD

**Issue by felixfontein** [ansible-community/antsibull-core#15]
@gotmax23 what do you think about the following?

1. Create `stable-1` branch, bump `main` to `2.0.0a1`
2. Drop Python 3.6 and 3.7 support, maybe even 3.8?
3. Adjust antsibull-docs and antsibull to work with both antsibull-core 1.x.y and 2.x.y
4. Once we're happy with the current state of antsibull-core, release 2.0.0

antsibull-docs will have to support antsibull-core 1.x.y for the remainder of its 1.x.y release cycle, but at least we can get rid of the Python < 3.8 compatiblity code in antsibull-core.

If we ever need a new feature (or bugfix) in antsibull-core 1.x.y for antsibull-docs, we can still add it (via the `stable-1` branch).

**Comment by felixfontein** [comment]
> > Drop Python 3.6 and 3.7 support, maybe even 3.8?
> 
> We only dropped support for 3.6 and 3.7 in antsibull. I'm fine with dropping 3.9, but we should be consistent.

I don't think it is necessary to fully align these; with Python 3.8 antsibull can still use antsibull-core 1.x.y, while with Python 3.9+ it can use antsibull-core 2.x.y.

> > Adjust antsibull-docs and antsibull to work with both antsibull-core 1.x.y and 2.x.y
> 
> antsibull-docs won't be able to work with antsibull-core 2 and still support 3.6/3.7 if we get rid of the compatibility code. 

Why should it do that? With Python 3.6/3.7/3.8 `pip install antsibull-docs` would install antsibull-core < 2.0.0, and with Python 3.9+ it would install antsibull-core >= 2.0.0.

> Also, how would this affect antsibull-changelog?

antsibull-changelog does not use antsibull-core.

**Comment by gotmax23** [comment]
> Why should it do that? With Python 3.6/3.7/3.8 pip install antsibull-docs would install antsibull-core < 2.0.0, and with Python 3.9+ it would install antsibull-core >= 2.0.0.

I'm not talking about pip. I'm talking about removing compatibility code from antsibull-core v2 that antsibull-docs depends on (e.g. `antsibull_core.compat.asyncio_run`). Or did you mean something else by "get rid of the Python < 3.8 compatiblity code in antsibull-core"?

**Comment by felixfontein** [comment]
Something like
```py
import asyncio

try:
    asyncio_run = asyncio.run
except AttributeError:
    from antsibull_core.compat import asyncio_run
```
should work fine.

**Comment by gotmax23** [comment]
Meh, I'd rather just change compat.py to something like (

``` python
import asyncio
import sys
from importlib import metadata

best_get_loop = asyncio.get_running_loop
asyncio_run = asyncio.run
create_task = asyncio.create_task


if sys.version_info < (3, 9):
    #: In Python3.9+, argparse.BooleanOptionalAction gives us a simple way to add
    #: --feature/--no-feature command line switches.  In earluer versions, we use
    #: code that we've copied from the upstream code.
    from .vendored._argparse_booleanoptionalaction import BooleanOptionalAction
else:
    # As long as https://bugs.python.org/issue46080 is not fixed, still use vendored version so
    # that users can actually use `--help` for affected subcommands. Once that is fixed and
    # included in released Python versions, we can adjust the sys.version_info check above.
    #
    #     from argparse import BooleanOptionalAction
    from .vendored._argparse_booleanoptionalaction import BooleanOptionalAction

__all__ = ('BooleanOptionalAction', 'asyncio_run', 'best_get_loop', 'create_task', 'metadata')
```

) and add a deprecation warning to the changelog that it'll be removed entirely in the next major release (i.e. 3.0.0). I'm not entirely opposed to your approach, but it adds unnecessary code duplication IMO. We need to keep the BooleanOptionalAction part, and I don't think it's a big deal to keep the other members without the conditional imports for old Pythons.

---

I think we should also add both versions of antsibull-core to the test matrix for antsibull and antsibull-docs. antsibull's test suite isn't super comprehensive. but at least we'll be able to check that the basic functionality works with both versions of core.

As for maintenance, we should explicitly lay out the policy. Are both versions fully supported? Which types of features will be backported?

Also, is it really worthwhile to do the make antsibull which does not need to support 3.6 and 3.7 work with both versions of core?  I'm fine with doing so, but I'm curious about your rationale.

Besides that, I'm broadly in support of this plan :).

**Comment by felixfontein** [comment]
> Meh, I'd rather just change compat.py to something like [...] and add a deprecation warning to the changelog that it'll be removed entirely in the next major release (i.e. 3.0.0).

That's actually better, let's do it that way.

> I think we should also add both versions of antsibull-core to the test matrix for antsibull and antsibull-docs. antsibull's test suite isn't super comprehensive. but at least we'll be able to check that the basic functionality works with both versions of core.

Yes, that's what I intend to do anyway.

> As for maintenance, we should explicitly lay out the policy. Are both versions fully supported? Which types of features will be backported?

Both versions are supported (for now), but features will only go to the `main` branch and will not be backported, except if we really need it for antsibull / antsibull-docs :) That's not really a great policy, but well...

> Also, is it really worthwhile to do the make antsibull which does not need to support 3.6 and 3.7 work with both versions of core? I'm fine with doing so, but I'm curious about your rationale.

Actually for antsibull we don't really need to do that - *except* if we drop Python 3.9 support here as well.

So I guess the big question is whether to drop 3.9 here as well or not. The advantage of dropping support for it is that we want to do new major versions very seldomly (both for antsibull-core and antsibull-docs, since they make upgrading antsibull-docs in ansible-core sanity tests / docs build more painful). antsibull itself can easily drop support for Python 3.9 in a new release, but antsibull-core and antsibull-docs cannot.

**Comment by gotmax23** [comment]
It'd be nice to align antsibull's supported python versions with antsibull-core v2's supported python versions and then tie antsibull to antsibull-core v2. That way, we'd only have to backport features for antsibull-docs.

**Comment by felixfontein** [comment]
It would definitely be nice, but I don't think have to do it at once. I think it's really better to also drop 3.8 from antsibull-core, and still keep supporting 3.8 for a bit longer in antsibull (but not too long, maybe only a few months?). Does that sound ok?

**Comment by gotmax23** [comment]
> It would definitely be nice, but I don't think have to do it at once. I think it's really better to also drop 3.8 from antsibull-core, and still keep supporting 3.8 for a bit longer in antsibull (but not too long, maybe only a few months?). Does that sound ok?

Sounds good to me; let's drop support for 3.6, 3.7, and 3.8 from antsibull-core now while we're doing a new major release.

**Comment by felixfontein** [comment]
I created a `stable-1` branch and began with a PR (#16) to bump the version of `main` to 2.0.0.

**Comment by felixfontein** [comment]
I'm currently planning of proceeding in this order:
1. Merge #18 (once reviewed)
2. Merge https://github.com/ansible-community/antsibull-docs/pull/81 and https://github.com/ansible-community/antsibull/pull/464
3. Assuming the corresponding CIs pass, release antsibull 0.52.0 and antsibull-docs 1.8.0.
4. Update CI of antsibull-core to run compatibility tests against 0.52.0 and 1.8.0 instead of against the respective main branches.
5. Wait a bit, see if we want to add more changes to antsibull-core 2.0.0.
6. Once we are done, release antsibull-core 2.0.0.

Does that sound good?

**Comment by gotmax23** [comment]
Sounds good to me. Originally, we discussed releasing an antsibull-core 2.0.0a1. I'll defer to you on whether that's still worth it or not :).

**Comment by gotmax23** [comment]
I'm weakly in favor of a 2.0.0a1 release

**Comment by felixfontein** [comment]
Hmm, looks like we never got further than the 2.0.0a1 release :) @gotmax23 do you mind if I do a 2.0.0 release now? Or do you think we should wait longer?

**Comment by felixfontein** [comment]
Let's release 2.0.0 next week. Does that sound good?

**Comment by gotmax23** [comment]
Sounds good to me

**Comment by felixfontein** [comment]
It's time for 2.0.0 :tada:
